### PR TITLE
Add missing template parameter to intermediate result

### PR DESCRIPTION
**Pull request by jppelteret**
Refers to #5829

### CODE REVIEW

**Comment by jppelteret** [comment]
Ah drats, I made a mistake and stuffed that up. Fix incoming...

**Comment by bangerth** [comment]
This fails:
```
/home/bob/source/include/deal.II/base/tensor.h:962:15: error: ambiguous overload for â€˜operator=â€™ (operand types are â€˜dealii::Tensor<0, 2, const double>â€™ and â€˜dealii::Tensor<0, 2, const double>â€™)
     values[i] = Tensor<rank_-1,dim,Number>(initializer[i]);
               ^
/home/bob/source/include/deal.II/base/tensor.h:962:15: note: candidates are:
/home/bob/source/include/deal.II/base/tensor.h:800:23: note: dealii::Tensor<0, dim, Number>& dealii::Tensor<0, dim, Number>::operator=(const dealii::Tensor<0, dim, OtherNumber>&) [with OtherNumber = const double; int dim = 2; Number = const double]
 Tensor<0,dim,Number> &Tensor<0,dim,Number>::operator = (const Tensor<0,dim,OtherNumber> &p)
                       ^
/home/bob/source/include/deal.II/base/tensor.h:810:23: note: dealii::Tensor<0, dim, Number>& dealii::Tensor<0, dim, Number>::operator=(const OtherNumber&) [with OtherNumber = dealii::Tensor<0, 2, const double>; int dim = 2; Number = const double]
 Tensor<0,dim,Number> &Tensor<0,dim,Number>::operator = (const OtherNumber &d)
                       ^
/home/bob/source/include/deal.II/base/tensor.h:87:7: note: dealii::Tensor<0, 2, const double>& dealii::Tensor<0, 2, const double>::operator=(const dealii::Tensor<0, 2, const double>&) <deleted>
 class Tensor<0,dim,Number>
       ^
```
I think https://github.com/dealii/dealii/commit/2263ac72b63a99089a3adcd55cde5172c1fce1a8 causes this ambiguity.

**Comment by bangerth** [comment]
We may have to have another assignment operator that takes `Number` and `Tensor<0,dim,Number>` since if the left and right hand sides of the assignment are the same, the compiler may have to choose between two operators that both are acceptable.

**Comment by jppelteret** [comment]
No, hold on... I just got the result types wrong. I've just fixed this locally...

**Comment by jppelteret** [comment]
@bangerth @davydden  Fixed (but waiting for the tester, obviously).

Of course, if this does nothing to mitigate #5829 then we should just close this and go the route that @QiaoLei-88 proposed.

**Comment by QiaoLei-88** [comment]
The same error but moved to different place
```
/home/qiaol/devel/dealii_git/include/deal.II/base/tensor.h(363): error: initial value of reference to non-const must be an lvalue
  class Tensor
        ^
          detected during:
            implicit generation of "dealii::Tensor<0, 2, Sacado::Rad::ADvar<double>> &dealii::Tensor<0, dim, Number>::operator=(dealii::Tensor<0, 2, Sacado::Rad::ADvar<double>> &&) [with dim=2, Number=Sacado::Rad::ADvar<double>]" at line 962
            instantiation of class "dealii::Tensor<0, dim, Number> [with dim=2, Number=Sacado::Rad::ADvar<double>]" at line 962
            instantiation of "dealii::Tensor<rank_, dim, Number>::Tensor(const dealii::Tensor<rank_, dim, OtherNumber> &) [with rank_=1, dim=2, Number=Sacado::Rad::ADvar<double>, OtherNumber=double]" at line 531 of "/home/qiaol/devel/dealii_git/source/fe/fe_values.cc"
            instantiation of "void dealii::FEValuesViews::internal::do_function_derivatives<order,dim,spacedim,Number>(const dealii::ArrayView<Number> &, const dealii::Table<2, dealii::Tensor<order, spacedim, double>> &, const std::vector<dealii::FEValuesViews::Scalar<dim, spacedim>::ShapeFunctionData, std::allocator<dealii::FEValuesViews::Scalar<dim, spacedim>::ShapeFunctionData>> &, std::vector<dealii::ProductType<Number, dealii::Tensor<order, spacedim, double>>::type,
                      std::allocator<dealii::ProductType<Number, dealii::Tensor<order, spacedim, double>>::type>> &) [with order=1, dim=1, spacedim=2, Number=const Sacado::Rad::ADvar<double>]" at line 1411 of "/home/qiaol/devel/dealii_git/source/fe/fe_values.cc"
            instantiation of
                      "void dealii::FEValuesViews::Scalar<dim, spacedim>::get_function_gradients_from_local_dof_values(const InputVector &, std::vector<dealii::FEValuesViews::Scalar<dim, spacedim>::OutputType<InputVector::value_type>::gradient_type, std::allocator<dealii::FEValuesViews::Scalar<dim, spacedim>::OutputType<InputVector::value_type>::gradient_type>> &) const [with dim=1, spacedim=2, InputVector=std::vector<Sacado::Rad::ADvar<double>, std::allocator<Sacado::Rad::ADvar<double>>>]" at
                      line 24727 of "/home/qiaol/package_install/deal.II-git/build_devel/source/fe/fe_values.inst"
```

More fixes needed.

**Comment by jppelteret** [comment]
@QiaoLei-88 Thanks for the update. So, in effect, that did nothing :-/ Then I think that we need not proceed along these lines. Thanks for trying this.

File: source/fe/fe_values.cc [hunk 1]
```diff
@@ -510,8 +510,8 @@ namespace FEValuesViews
                                                shape_derivatives[0].size() : derivatives.size();
       AssertDimension (derivatives.size(), n_quadrature_points);
 
-      std::fill (derivatives.begin(), derivatives.end(),
-                 typename ProductType<Number,dealii::Tensor<order,spacedim> >::type());
+      typedef typename ProductType<Number,dealii::Tensor<order,spacedim> >::type ProductTensorType;
```
Comment (drwells) [inline source/fe/fe_values.cc]: why not just use `typename derivatives::value_type`?

Comment (jppelteret) [reply source/fe/fe_values.cc]: True, that would definitely be more concise. If we choose to follow through with this PR then I'll change this first.

**Comment by drwells** [comment]
@jppelteret  It looks like I should have read the whole conversation before reviewing the typedef changes.

**Comment by jppelteret** [comment]
I realise that this is not only __not__ a solution to the current problem, but it actually is counterproductive when using AD numbers (it will create additional temporary entries). Furthermore, we specifically [generalised the `Tensor::operator*`](https://github.com/dealii/dealii/blob/master/include/deal.II/base/tensor.h#L1517) to deal with the scalar, tensor and result all having different number types.

**Comment by jppelteret** [comment]
@drwells No worries, and thank you for your input regardless of the outcome of the PR :-)

**Pull request closed without merging**